src/wg/image/imagebuffer.d

```diff
@@ -24,6 +28,68 @@ struct ImageBuffer
 }
 
 
+/**
+ * Allocate an image buffer for the specified format in GC memory.
+ * Optionally allocate additional metadata pages.
+ */
+ImageBuffer allocImage(MetadataBlocks...)(const(char)[] format, uint width, uint height, size_t[] additionalMetadataBytes...)
+{
+    import std.exception : enforce;
+
+    assert(additionalMetadataBytes.length == MetadataBlocks.length);
+
+    ImageBuffer img;
+    void[] mem = allocImageImpl(format, true, width, height, getMetadataSize!(false, MetadataBlocks)(additionalMetadataBytes), getGcAllocator(), img);
+    enforce(mem.length > 0, "Failed to allocate image with format: " ~ format);
+    arrangeMetadata!(false, MetadataBlocks)(img.metadata, additionalMetadataBytes);
+    return img;
+}
+
+/**
+ * Allocate an image buffer for the specified format using the supplied allocator.
+ * Optionally allocate additional metadata pages.
+ */
+ImageBuffer allocImage(MetadataBlocks...)(const(char)[] format, uint width, uint height, Allocator* allocator, size_t[] additionalMetadataBytes...) nothrow @nogc
+{
+    assert(additionalMetadataBytes.length == MetadataBlocks.length);
+
+    ImageBuffer img;
+    void[] mem = allocImageImpl(format, true, width, height, getMetadataSize!(true, MetadataBlocks)(additionalMetadataBytes), allocator, img);
+    if (mem.length)
+    {
+        AllocationMetadata* allocData = arrangeMetadata!(true, MetadataBlocks)(img.metadata, additionalMetadataBytes);
+        allocData.allocations[0].allocator = allocator;
+        allocData.allocations[0].mem = mem;
+    }
+    return img;
+}
+
+/**
+ * Free any allocations contained in an image buffer.
+ */
+void freeImage(ref ImageBuffer image) nothrow @nogc
+{
+    AllocationMetadata* allocData = image.getMetadata!AllocationMetadata();
+    if (allocData)
+    {
+        // cache the allocation pages locally, since we will probably delete them along the way
+        AllocationMetadata.Page[128] allocations;
+        allocations[0 .. allocData.allocations.length] = allocData.allocations[];
```

Comment:
Actually, this should be okay no? It will range error if `allocData.allocations.length > allocations.length`. The assert would just repeat the bounds checker test?

Reply:
hmm, yes, it will repeat the bounds checker test in fact. I would use something like `std.range.chunk` to iterate over `allocData.allocation` to free pages safely but if `phobos` is unwanted then it just can be improved later, you're right bounds checker detects violation.

Reply:
How would `chunks` help? It need to copy the whole thing... the list of allocations is one of those allocations, and it's really not worth detecting which one to free it last.

Reply:
Also, the number of metadata pages is like to be in the order of 2. 128 is a crazy over-estimate. It'll be fine :) .. it's best to use the stack for temp memory, that's what it's for. I could implement a fail-over allocation thought for `@safe`-ty, but I feel a runtime assert is also @safe...

Reply:
I prefer stack allocation too, hardcoded size were confusing me. But current solution is good enough. It's just not important "issue" and can be easily improved later.